network/src/peer_manager/mod.rs

```diff
@@ -459,50 +459,63 @@ where
+                            .get(&conn.metadata.remote_peer_id)
+                            .map_or(PeerRole::Unknown, |auth_context| auth_context.role);
+                    }
+                    ConnectionOrigin::Inbound => {
+                        // Everything below here is meant for unknown peers only, role comes from
+                        // Noise handshake and if it's not `Unknown` it is trusted
+                        if conn.metadata.role != PeerRole::Unknown {
+                            return;
+                        }
+                        // TODO: Keep track of somewhere else to not take this hit in case of DDoS
+                        // Count unknown inbound connections
+                        let unknown_inbound_conns = self
+                            .active_peers
+                            .iter()
+                            .filter(|(peer_id, (metadata, _))| {
+                                metadata.origin == ConnectionOrigin::Inbound
+                                    && trusted_peers
+                                        .get(peer_id)
+                                        .map_or(true, |peer| peer.role == PeerRole::Unknown)
+                            })
+                            .count();
+
+                        // Reject excessive inbound connections made by unknown peers
+                        // We control outbound connections with Connectivity manager before we even send them
+                        // and we must allow connections that already exist to pass through tie breaking.
+                        if self
+                            .active_peers
+                            .contains_key(&conn.metadata.remote_peer_id)
+                            || unknown_inbound_conns < self.inbound_connection_limit
+                        {
+                            info!(
+                                NetworkSchema::new(&self.network_context)
+                                    .connection_metadata_with_address(&conn.metadata),
+                                "{} New connection established: {}",
+                                self.network_context,
+                                conn.metadata
+                            );
+                            // Add new peer, updating counters and all
+                            self.add_peer(conn);
+                            self.update_connected_peers_metrics();
+                        } else {
+                            info!(
+                                NetworkSchema::new(&self.network_context)
+                                    .connection_metadata_with_address(&conn.metadata),
+                                "{} Connection rejected due to connection limit: {}",
+                                self.network_context,
+                                conn.metadata
+                            );
+                            self.disconnect(conn);
+                        }
+                    }
```

Comment:
Taking a step back here for a moment, does it actually make sense to allow even trusted peers infinite inbound connections? Trusted peers can also be compromised, run with buggy k8s auto scaling routines, etc.... We allow unbounded outbound connections here because it's something we initiate and control (and we already bound it in connmgr); inbound connections are outside of our control and so they need to be bounded.

It makes more sense to me to have a min priority queue like `BTreeMap<PeerRole, PeerId>` that we can pop-min from when we're at the connection limit and need to evict a connection.

Reply:
So, I discussed this earlier when I initially added the inbound connection limit.

As an MVP, we can accept all the trusted connections as they're a limited set (# in config + # in onchain discovery).  The other option would yes, to be feed through the `PeerRole` stratification of trusted peers, but realistically like I said there's a limited set for now, and we can always come back to add that kind of information